### PR TITLE
[custom_ops] expose torch.library.register_torch_dispatch

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130261
* #130064

This is the API for defining the interaction between a torch_dispatch
class and a custom op. Taking API bikeshedding.

Test Plan:
- new tests

### CODE REVIEW

File: torch/library.py [hunk 1]
```diff
@@ -870,6 +871,84 @@ def register_autograd(
...
+        >>> @torch.library.custom_op("mylib::foo", mutates_args={})
+        >>> def foo(x: torch.Tensor) -> torch.Tensor:
+        >>>     return x.clone()
+        >>>
+        >>> class MyMode(torch.utils._python_dispatch.TorchDispatchMode):
+        >>>     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        >>>         return func(*args, **kwargs)
+        >>>
+        >>> @torch.library.register_torch_dispatch("mylib::foo", MyMode)
+        >>> def _(mode, func, types, args, kwargs):
```
Comment (zou3519) [inline torch/library.py:910]: TODO: add a section to extending.rst and link it from here

Comment (zou3519) [reply torch/library.py:913]: done

File: docs/source/notes/extending.rst [hunk 2]
```diff
@@ -914,6 +914,24 @@ It is also possible to add `new` native functions using :mod:`torch.library`. Th
...
+.. code-block:: python
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        pass
+
+When a user calls an operator with inputs that have ``__torch_dispatch__``, that call
+may be forwarded to the ``__torch_dispatch__``. args and kwargs get normalized before
+the call to ``__torch_dispatch__``, that is:
+- the ``kwargs`` consist of keyword-only arguments in the operator's schema
```
Comment (albanD) [inline docs/source/notes/extending.rst:931]:
```suggestion
- the ``kwargs`` consist of keyword-only arguments in the operator's schema, it is only passed if it is not equal to the default value provided in the schema.
```

Comment (albanD) [reply docs/source/notes/extending.rst:931]: Or something similar.

File: docs/source/notes/extending.rst [hunk 3]
```diff
@@ -914,6 +914,24 @@ It is also possible to add `new` native functions using :mod:`torch.library`. Th
...
+- the ``args`` consists of all other arguments, no matter how they were called
```
Comment (albanD) [inline docs/source/notes/extending.rst:932]:
```suggestion
- the ``args`` consists of all other arguments, no matter how they were passed to the op
```

**Review by albanD — APPROVED** [review]
Sounds good!
Only small nits on the doc

**Comment by albanD** [comment]
Ho we also remove positional? lol good to know!

**Comment by zou3519** [comment]
> Ho we also remove positional? lol good to know!

Yes, I'm not a big fan lol

**Pull request closed without merging**